### PR TITLE
AE-452 UMA admin events

### PR DESCRIPTION
**Pull request by arbitrarylink**
UMA Voting Contract - Event Monitoring
Contract: 0x8B1631ab830d11531aE83725fDa4D86012eCCd77
Events listed in the contract ABI:
OwnershipTransferred
EncryptedVote
PriceRequestAdded
PriceResolved
RewardsRetrieved
VoteCommitted
VoteRevealed

Agent actions:
Generate a finding for each of these events
OwnershipTransferred should be treated as Suspicious+High severity

UMA Governance Contract - Event Monitoring
Contract: 0x592349F7DeDB2b75f9d4F194d4b7C16D82E507Dc
Events listed in the contract ABI:
NewProposal
ProposalExecuted
AddedSharedMember
RemovedSharedMember
ResetExclusiveMember

Agent actions:
Generate a finding for each of these events
ResetExclusiveMember should be considered Suspicious 

From the UMA team:  “I think the regular voting methods (like EncryptedVote) can probably be dropped entirely as they are pretty routine, unless routine events typically produce reports in Forta.”

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -27,6 +27,12 @@ the following handlers:
   - Type is always set to "suspicious"
   - Metadata field contains to and from addresses
 
+- AE-UMA-ADMIN-EVENT
+  - Fired on any event in admin-events.json
+  - Severity is set to value in admin-events.json
+  - Type is set to value in admin-events.json
+  - Metadata field contains Contract Name, Contract Address and Event Name
```
Comment (echoAE) [inline README.md:34]:
```suggestion
  - Metadata field contains contract name, contract address, and event name
```

File: src/admin-events/admin-events.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
+const { Finding, FindingSeverity, FindingType } = require('forta-agent');
+const { getAbi, getAddress } = require('@uma/contracts-node');
+const ethers = require('ethers');
+
+// load config files
+const config = require('../../agent-config.json');
```
Comment (echoAE) [inline src/admin-events/admin-events.js:6]: Optional: You are only interested in the `umaEverestId` value from the `agent-config.json` file, so you could use destructuring here:
```suggestion
const { umaEverestId } = require('../../agent-config.json');
```
Then line 56 would change to:
```
    everestId: umaEverestId
```

File: src/admin-events/admin-events.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  });
+}
+
+async function handleTransaction(txEvent) {
+  const findings = [];
+
+  // iterate over each contract name to get the address and events
+  contractNames.forEach(async (contractName) => {
+    // for each contract name, lookup the address, events and interface
+    const chainId = 1;
```
Comment (echoAE) [inline src/admin-events/admin-events.js:71]: I favor putting this variable assignment outside the handler function, though that might be a matter of personal preference.

File: src/admin-events/admin-events.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    const contractAddressPromise = getAddress(contractName, chainId);
+    const contractAddress = (await contractAddressPromise).toLowerCase();
```
Comment (echoAE) [inline src/admin-events/admin-events.js:73]: The variable `contractAddressPromise` is not used anywhere else.  I wouldn't be upset if you eliminate line 72 and modify line 73:
```suggestion
    const contractAddress = (await getAddress(contractName, chainId)).toLowerCase();
```

File: src/admin-events/admin-events.spec.js [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+// Load Config files
+const {
+  TransactionEvent,
+  FindingType,
+  FindingSeverity,
+  Finding,
+} = require('forta-agent');
+
+const config = require('../../agent-config.json');
```
Comment (echoAE) [inline src/admin-events/admin-events.spec.js:12]: Same comment from the handler module:
```suggestion
const { umaEverestId } = require('../../agent-config.json');
```

File: src/admin-events/admin-events.spec.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+const chainId = 1;
+const votingAddressPromise = getAddress('Voting', chainId);
+
+const { handleTransaction } = require('./admin-events');
+
+/**
+ * TransactionEvent(type, network, transaction, receipt, traces, addresses, block)
+ */
+
+// Can this be imported from Forta Library?
```
Comment (echoAE) [inline src/admin-events/admin-events.spec.js:23]: Note: Resolve the question in this comment, then delete the comment.

File: src/admin-events/admin-events.spec.js [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+          description: `The ${eventName} event was emitted by the ${contractName} contract`,
+          alertId: 'AE-UMA-ADMIN-EVENT',
+          type: FindingType.Unknown,
+          severity: FindingSeverity.Low,
+          metadata: {
+            contractName,
+            contractAddress,
+            eventName,
+          },
+          everestId: config.umaEverestId,
```
Comment (echoAE) [inline src/admin-events/admin-events.spec.js:127]: See comment above about using destructuring with the `require('../../agent-config.json')` statement.

File: src/admin-events/admin-events.spec.js [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        logs: logsMatchEvent,
+        addresses: { [votingContract]: true },
+      });
+
+      // run agent
+      const findings = await handleTransaction(txEvent);
+
+      // assertions
+      expect(findings).toStrictEqual([
+        Finding.fromObject({
```
Comment (echoAE) [inline src/admin-events/admin-events.spec.js:116]: Consider importing the `createAlert` function from the handler module and using it with your mocked variables to create a `Finding` object.

**Review by echoAE — CHANGES_REQUESTED** [review]
Relatively minor suggestions and comments.

File: README.md (hunk 1 above)
Comment (arbitrarylink) [reply README.md]: Done

File: src/admin-events/admin-events.js (hunk 2 above)
Comment (arbitrarylink) [reply src/admin-events/admin-events.js]: Done

File: src/admin-events/admin-events.js (hunk 3 above)
Comment (arbitrarylink) [reply src/admin-events/admin-events.js]: Done

File: src/admin-events/admin-events.js (hunk 4 above)
Comment (arbitrarylink) [reply src/admin-events/admin-events.js]: Done

File: src/admin-events/admin-events.spec.js (hunk 5 above)
Comment (arbitrarylink) [reply src/admin-events/admin-events.spec.js]: Done

File: src/admin-events/admin-events.spec.js (hunk 7 above)
Comment (arbitrarylink) [reply src/admin-events/admin-events.spec.js]: Done

File: src/admin-events/admin-events.js [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+// helper function to create alerts
+function createAlert(log, contractName, contractAddress, eventType, eventSeverity) {
+  const eventName = log.name;
+  return Finding.fromObject({
+    name: 'UMA Admin Event',
+    description: `The ${eventName} event was emitted by the ${contractName} contract`,
+    alertId: 'AE-UMA-ADMIN-EVENT',
+    type: FindingType[eventType],
+    severity: FindingSeverity[eventSeverity],
+    everestId: config.umaEverestId,
```
[the remark this answers is not in the document]
Comment (arbitrarylink) [reply src/admin-events/admin-events.js]: Done

File: src/admin-events/admin-events.spec.js (hunk 8 above)
Comment (arbitrarylink) [reply src/admin-events/admin-events.spec.js]: Done

File: src/admin-events/admin-events.spec.js (hunk 6 above)
Comment (arbitrarylink) [reply src/admin-events/admin-events.spec.js]: Done

**Review by arbitrary-cr0 — APPROVED** [review]

File: src/admin-events/admin-events.js [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+  const iface = new ethers.utils.Interface(abi);
+
+  // Create an association between the contract name and the interface
+  ifaces[contractName] = iface;
+});
+
+// Filters the logs to only events in eventNames
+function filterAndParseLogs(logs, address, iface, eventNames) {
+  // collect logs only from the contract
+  const contractLogs = logs.filter((log) => log.address === address);
```
Comment (arbitrary-kdawg) [inline src/admin-events/admin-events.js:37]: Seems like the function shouldn't continue past this point if the address doesn't match--I would return from the function  immediately if contractLogs is empty.

File: src/admin-events/admin-events.js [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+      contractName,
+      contractAddress,
+      eventName,
+    },
+  });
+}
+
+async function handleTransaction(txEvent) {
+  const findings = [];
+  const chainId = 1;
```
Comment (arbitrary-kdawg) [inline src/admin-events/admin-events.js:67]: Move the chainId constant out of the function?
Also should this be CHAIN_ID according to the Google JS style?

File: src/admin-events/admin-events.json [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+    "PriceRequestAdded" : {"type" : "Unknown", "severity" : "Low"},
+    "PriceResolved" : {"type" : "Unknown", "severity" : "Low"},
+    "RewardsRetrieved" : {"type" : "Unknown", "severity" : "Low"},
+    "VoteCommitted" : {"type" : "Unknown", "severity" : "Low"},
+    "VoteRevealed" : {"type" : "Unknown", "severity" : "Low"}
+  },
+  "Governor" : {
+    "NewProposal" : {"type" : "Unknown", "severity" : "Low"},
+    "ProposalExecuted" : {"type" : "Unknown", "severity" : "Low"},
+    "AddedSharedMember" : {"type" : "Unknown", "severity" : "Low"},
```
Comment (arbitrary-kdawg) [inline src/admin-events/admin-events.json:13]: Maybe AddedSharedMember and RemovedSharedMember should be promoted to type "suspicious"?

File: src/admin-events/admin-events.spec.js [hunk 13]
```diff
@@ -0,0 +1,106 @@
+const ethers = require('ethers');
+const { getAddress } = require('@uma/contracts-node');
+
+// Load Config files
```
Comment (arbitrary-kdawg) [inline src/admin-events/admin-events.spec.js:4]: this "Load Config Files" comment doesn't match the code

File: src/admin-events/admin-events.spec.js [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+const { TransactionEvent } = require('forta-agent');
+
+const chainId = 1;
+const votingAddressPromise = getAddress('Voting', chainId);
+
+const { createAlert, handleTransaction } = require('./admin-events');
```
Comment (arbitrary-kdawg) [inline src/admin-events/admin-events.spec.js:10]: I'd group all of the require statements together at the top of the file

File: src/admin-events/admin-events.spec.js [hunk 15]
```diff
@@ -0,0 +1,106 @@
+const ethers = require('ethers');
+const { getAddress } = require('@uma/contracts-node');
+
+// Load Config files
+const { TransactionEvent } = require('forta-agent');
+
+const chainId = 1;
```
Comment (arbitrary-kdawg) [inline src/admin-events/admin-events.spec.js:7]: Same comment from admin-events.js -- we haven't been consistent about this, but chainId should probably be CHAIN_ID

File: src/admin-events/admin-events.spec.js [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+const votingAddressPromise = getAddress('Voting', chainId);
+
+const { createAlert, handleTransaction } = require('./admin-events');
+
+// TransactionEvent(type, network, transaction, receipt, traces, addresses, block)
+function createTxEvent({ logs, addresses }) {
```
Comment (arbitrary-kdawg) [inline src/admin-events/admin-events.spec.js:13]: forta-agent provides a createTransactionEvent function--it's preferable to use that unless there's a reason not to

File: src/admin-events/admin-events.spec.js [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+    it('returns empty findings if contract address matches but not event', async () => {
+      const votingContract = (await votingAddressPromise).toLowerCase();
+
+      // logs data for test case: address match + no topic match
+      const logsNoMatchEvent = [
+        {
+          address: votingContract,
+          topics: [
+            ethers.utils.keccak256(ethers.utils.toUtf8Bytes('EncryptedVote(address,uint256,bytes32,uint256,bytes,bytes)')),
+            ethers.constants.HashZero,
```
Comment (arbitrary-kdawg) [inline src/admin-events/admin-events.spec.js:53]: nit: I like to add comments to identify the topics just so it's clear what we're assigning HashZero to
e.g. for EncryptedVote:
ethers.constants.HashZero, // voter
ethers.constants.HashZero, // roundId
ethers.constants.HashZero, // identifier

File: src/admin-events/admin-events.spec.js [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+            ethers.constants.HashZero,
+            ethers.constants.HashZero,
+          ],
+          data: `0x${'0'.repeat(1000)}`,
```
Comment (arbitrary-kdawg) [inline src/admin-events/admin-events.spec.js:57]: To make it clear what you're doing, maybe add a comment here for 'data'
e.g.
// create a large dummy array to give ethers.parseLog() something to decode

File: src/admin-events/admin-events.spec.js [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+      const contractName = 'Voting';
+      const contractAddress = votingContract;
+
+      // logs data for test case: address match + topic match (should trigger a finding)
+      const logsMatchEvent = [
+        {
+          address: votingContract,
+          topics: [
+            ethers.utils.keccak256(ethers.utils.toUtf8Bytes('VoteCommitted(address,uint256,bytes32,uint256,bytes)')),
+            ethers.constants.HashZero,
```
Comment (arbitrary-kdawg) [inline src/admin-events/admin-events.spec.js:86]: ethers.constants.HashZero, // voter
ethers.constants.HashZero, // roundId
ethers.constants.HashZero, // identifier

File: src/admin-events/admin-events.spec.js [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+            ethers.constants.HashZero,
+            ethers.constants.HashZero,
+          ],
+          data: `0x${'0'.repeat(1000)}`,
```
Comment (arbitrary-kdawg) [inline src/admin-events/admin-events.spec.js:90]: add same explanatory comment for 'data' as above

File: package.json [hunk 21]
```diff
@@ -14,6 +14,7 @@
     "test": "jest"
   },
   "dependencies": {
+    "@uma/contracts-node": "^0.1.0",
     "bignumber.js": "^9.0.1",
     "ethers": "^5.4.6",
     "forta-agent": "^0.0.15",
```
Comment (arbitrary-kdawg) [inline package.json:20]: forta-agent 0.0.16 was released on 9/24

File: src/admin-events/admin-events.js (hunk 11 above)
Comment (arbitrarylink) [reply src/admin-events/admin-events.js]: Done

File: src/admin-events/admin-events.json (hunk 12 above)
Comment (arbitrarylink) [reply src/admin-events/admin-events.json]: Done

File: src/admin-events/admin-events.spec.js (hunk 13 above)
Comment (arbitrarylink) [reply src/admin-events/admin-events.spec.js]: Done

File: src/admin-events/admin-events.spec.js (hunk 14 above)
Comment (arbitrarylink) [reply src/admin-events/admin-events.spec.js:5]: Done

File: src/admin-events/admin-events.spec.js (hunk 15 above)
Comment (arbitrarylink) [reply src/admin-events/admin-events.spec.js]: Done

File: src/admin-events/admin-events.spec.js (hunk 17 above)
Comment (arbitrarylink) [reply src/admin-events/admin-events.spec.js]: Done

File: src/admin-events/admin-events.spec.js (hunk 18 above)
Comment (arbitrarylink) [reply src/admin-events/admin-events.spec.js:56]: Done

File: src/admin-events/admin-events.spec.js (hunk 19 above)
Comment (arbitrarylink) [reply src/admin-events/admin-events.spec.js]: Done

File: src/admin-events/admin-events.spec.js (hunk 20 above)
Comment (arbitrarylink) [reply src/admin-events/admin-events.spec.js:90]: Done

File: package.json (hunk 21 above)
Comment (arbitrarylink) [reply package.json]: Done

File: src/admin-events/admin-events.js (hunk 10 above)
Comment (arbitrarylink) [reply src/admin-events/admin-events.js:40]: Done

File: src/admin-events/admin-events.spec.js (hunk 16 above)
Comment (arbitrarylink) [reply src/admin-events/admin-events.spec.js]: Done

**Review by echoAE — APPROVED** [review]

**Pull request merged**